src/charm.py

```diff
@@ -582,26 +597,98 @@ def _wp_is_installed(self):
         logger.debug("Check if WordPress is installed")
         return self._run_wp_cli(["wp", "core", "is-installed"]).return_code == 0
 
-    def _current_effective_db_info(self):
+    @property
+    def _config_database_config(self) -> types_.DatabaseConfig | None:
+        """Database configuration details from charm config.
+
+        Returns:
+            Database configuration required to establish database connection.
+            None if not exists.
+        """
+        if any(
+            self.model.config.get(key) for key in ["db_host", "db_name", "db_user", "db_password"]
+        ):
+            return types_.DatabaseConfig(
+                hostname=self.model.config.get("db_host"),
+                database=self.model.config.get("db_name"),
+                username=self.model.config.get("db_user"),
+                password=self.model.config.get("db_password"),
+            )
+        return None
+
+    @property
+    def _db_relation_database_config(self) -> types_.DatabaseConfig | None:
+        """Database configuration details from stored state.
+
+        Since the legacy db relation is per-unit basis, the relation data must be stored in charm
```

Comment:
To help me understand, where in this function is the interaction with `StoredState`?

Reply:
It's under the function `_on_relation_db_changed`, where `MySQLDatabaseChangedEvent` will store the values in stored state. I've written the reasoning inside the docstring for the `_on_relation_db_changed` function as well, which in summary is that there's no way for the charm to select which database endpoint to connect to when accessing the unit ip(set) from the databag :)